### PR TITLE
Poetry error message

### PR DESCRIPTION
**Pull request by prathameshzarkar9**
Issue #845 

infoOutput has been updated and logged the message of the command if the command fails inside container

image

### CODE REVIEW

File: src/spec-common/injectHeadless.ts
Comment (samruddhikhandale) [inline src/spec-common/injectHeadless.ts:539]: Curious, why don't we always output the `err.cmdOutput` when there exists one ? That way even if the error code is not 127, everyone would benefit.

File: src/test/configs/dockerfile-with-target/.devcontainer.json
Comment (samruddhikhandale) [inline src/test/configs/dockerfile-with-target/.devcontainer.json:11]: Shouldn't we still keep this `"touch /tmp/postStartCommand.testmarker` ?

File: src/test/configs/example/.devcontainer.json
Comment (samruddhikhandale) [inline src/test/configs/example/.devcontainer.json:11]: Why did we update this? The tests using this config would fail as python Feature doesn't install poetry.

File: src/test/configs/dockerfile-with-target/.devcontainer.json
Comment (samruddhikhandale) [inline src/test/configs/dockerfile-with-target/.devcontainer.json:8]: Can we fix the failing tests?

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/spec-common/injectHeadless.ts
Comment (prathameshzarkar9) [reply src/spec-common/injectHeadless.ts:539]: done

File: src/test/configs/example/.devcontainer.json
Comment (prathameshzarkar9) [reply src/test/configs/example/.devcontainer.json:11]: reverted

File: src/test/configs/dockerfile-with-target/.devcontainer.json
Comment (prathameshzarkar9) [reply src/test/configs/dockerfile-with-target/.devcontainer.json:11]: reverted

File: src/test/configs/dockerfile-with-target/.devcontainer.json
Comment (prathameshzarkar9) [reply src/test/configs/dockerfile-with-target/.devcontainer.json]: fixed

File: src/spec-common/injectHeadless.ts
Comment (samruddhikhandale) [inline src/spec-common/injectHeadless.ts:538]: Should we check if `err.cmdOutput` exists before we output? We should try to avoid null refs.

File: src/test/cli.up.test.ts
Comment (samruddhikhandale) [inline src/test/cli.up.test.ts:28]: What's the purpose of this test with respect to the PR? 

If you want to test the poetry scenario, then you can create a new dev container config which fails similar to https://github.com/devcontainers/cli/issues/845

Then, validate if you the `poetry: command not found` error. However, in this case it's alright to not add the tests.

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/spec-common/injectHeadless.ts
Comment (prathameshzarkar9) [reply src/spec-common/injectHeadless.ts:538]: done

File: src/test/cli.up.test.ts
Comment (prathameshzarkar9) [reply src/test/cli.up.test.ts:28]: test removed

File: src/spec-common/injectHeadless.ts
Comment (samruddhikhandale) [inline src/spec-common/injectHeadless.ts:538]:
```suggestion
				if (err?.cmdOutput) {
```

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]
Let's wait for another review from @chrmarti. Thanks!

File: src/spec-common/injectHeadless.ts
Comment (chrmarti) [inline src/spec-common/injectHeadless.ts:538]: Should we move the error handling into `runSingleCommand` above, so it is done for each command? We might have missed that when adding the object syntax for user commands.

`runSingleCommand` uses `const printMode = name ? 'off' : 'continuous';`, so `cmdOutput` should already make it to the terminal when there are no names with the string/string-array syntax. Could you check?

**Review by chrmarti — CHANGES_REQUESTED** [review]
Left a comment, maybe the error handling should move up a few lines. Thanks.

File: src/spec-common/injectHeadless.ts
Comment (prathameshzarkar9) [reply src/spec-common/injectHeadless.ts:538]: The event is thrown in runRemoteCommand function, so I have added a cmdOutput to the message to show what exactly is failed in the command.
please check the new changes committed to fix this issue.

Comment (prathameshzarkar9) [reply src/spec-common/injectHeadless.ts:538]: moved the cmdOutput to runRemoteCommand function and appended in the message.

Comment (chrmarti) [reply src/spec-common/injectHeadless.ts]: I suggest to keep `message` of the error short and single line. Would it instead make sense to put https://github.com/devcontainers/cli/blob/main/src/spec-common/injectHeadless.ts#L506-L517 in a try-catch block and log `cmdOutput` in the catch? Much or all of the outer catch block's code then needs to move to the inner catch: https://github.com/devcontainers/cli/blob/main/src/spec-common/injectHeadless.ts#L533-L548.

I think the 'progress' events should move out of the `runSingleCommand` function. They look like they should only be sent once for all commands of a lifecycle hook. (This is confirmed by code in the Dev Containers extension.)

Comment (prathameshzarkar9) [reply src/spec-common/injectHeadless.ts]: Hi @chrmarti ,
I have checked in new changes as per your suggestion. Kindly help to review it and let me know if any change is required will accordingly address the same.

File: src/spec-common/injectHeadless.ts
Comment (chrmarti) [inline src/spec-common/injectHeadless.ts:549]: Keep this in a try-catch and move all infoOutput.event() calls here. These events don't seem to be per command but per lifecycle hook.

File: src/spec-common/injectHeadless.ts
Comment (chrmarti) [inline src/spec-common/injectHeadless.ts:521]: Only write the output when `name` is set. Otherwise the output was already written (see `printMode` above).

**Review by chrmarti — CHANGES_REQUESTED** [review]
Thanks @prathameshzarkar9! I have left a few comments.

File: src/spec-common/injectHeadless.ts
Comment (prathameshzarkar9) [reply src/spec-common/injectHeadless.ts:521]: Done

**Comment by prathameshzarkar9** [comment]
@chrmarti , can you please review the requested changes and approve the same.

File: src/spec-common/injectHeadless.ts
Comment (chrmarti) [inline src/spec-common/injectHeadless.ts:490]: Move this event to right before `await Promise.all(commands);`.

File: src/spec-common/injectHeadless.ts
Comment (chrmarti) [inline src/spec-common/injectHeadless.ts:510]: This is still missing an additional check. Maybe:
```suggestion
				if (printMode === 'off' && err?.cmdOutput) {
```

File: src/spec-common/injectHeadless.ts
Comment (chrmarti) [inline src/spec-common/injectHeadless.ts:539]: This event needs to move to right after `await Promise.all(commands);`.

File: src/spec-common/injectHeadless.ts
Comment (chrmarti) [inline src/spec-common/injectHeadless.ts:514]:
```suggestion
					infoOutput.raw(`\r\n\x1b[1m${name ? `${name} of ${lifecycleHookName}` : lifecycleHookName} from ${userCommandOrigin} interrupted.\x1b[0m\r\n\r\n`);
```

File: src/spec-common/injectHeadless.ts
Comment (chrmarti) [inline src/spec-common/injectHeadless.ts:517]:
```suggestion
						infoOutput.write(toErrorText(`${name ? `${name} of ${lifecycleHookName}` : lifecycleHookName} from ${userCommandOrigin} failed with exit code ${err.code}. Skipping any further user-provided commands.`));
```

File: src/spec-common/injectHeadless.ts
Comment (chrmarti) [inline src/spec-common/injectHeadless.ts:520]:
```suggestion
						description: `${name ? `${name} of ${lifecycleHookName}` : lifecycleHookName} from ${userCommandOrigin} failed.`,
```

File: src/spec-common/injectHeadless.ts
Comment (chrmarti) [inline src/spec-common/injectHeadless.ts:507]:
```suggestion
					infoOutput.raw(`\x1b[1mRunning ${name} of ${lifecycleHookName} from ${userCommandOrigin}...\x1b[0m\r\n${cmdOutput}\r\n`);
```

**Review by chrmarti — CHANGES_REQUESTED** [review]
A few more comments and some message polishing.

File: src/spec-common/injectHeadless.ts
Comment (prathameshzarkar9) [reply src/spec-common/injectHeadless.ts:514]: Done

File: src/spec-common/injectHeadless.ts
Comment (prathameshzarkar9) [reply src/spec-common/injectHeadless.ts:517]: Done

File: src/spec-common/injectHeadless.ts
Comment (prathameshzarkar9) [reply src/spec-common/injectHeadless.ts:520]: Done

File: src/spec-common/injectHeadless.ts
Comment (prathameshzarkar9) [reply src/spec-common/injectHeadless.ts:493]: done. moved this before `await Promise.all(commands);`

File: src/spec-common/injectHeadless.ts
Comment (prathameshzarkar9) [reply src/spec-common/injectHeadless.ts:542]: Done.

File: src/spec-common/injectHeadless.ts
Comment (chrmarti) [reply src/spec-common/injectHeadless.ts:493]: Actually, since `progressDetail` includes the command to run, maybe this should stay here.

@samruddhikhandale Could you advise on this? I think Codespaces is using the progress events in the UI?

@prathameshzarkar9 I don't see your latest changes on GitHub, did you push the commits?

Comment (prathameshzarkar9) [reply src/spec-common/injectHeadless.ts:493]: Hi @chrmarti ,
changes suggested by you are incorporated and the new changes also checked-in here in the latest commits.
PR is now updated can you please review it once @samruddhikhandale ?

Comment (samruddhikhandale) [reply src/spec-common/injectHeadless.ts:493]: > Actually, since progressDetail includes the command to run, maybe this should stay here.
> @samruddhikhandale Could you advise on this? I think Codespaces is using the progress events in the UI?

@chrmarti  Good catch here! 
Yes, Codespaces is still using `progressDetail` so let's keep it as is.

Comment (prathameshzarkar9) [reply src/spec-common/injectHeadless.ts:493]: @samruddhikhandale , @chrmarti 
I kept the progressDetails as it is. can you review it once?

**Comment by samruddhikhandale** [comment]
@prathameshzarkar9 Can you fix the failing tests? Thanks

File: src/spec-common/injectHeadless.ts
Comment (samruddhikhandale) [inline src/spec-common/injectHeadless.ts:547]: @prathameshzarkar9 Why was this added?

**Comment by Mathiyarasy** [comment]
@chrmarti Could you please review this PR ..

File: src/spec-common/injectHeadless.ts
Comment (chrmarti) [inline src/spec-common/injectHeadless.ts:543]: I think the error needs to be rethrown after reporting progress to preserve the previous behavior. Only if it is rethrown will the "further user-provided commands" be skipped. Please check.

File: src/spec-common/injectHeadless.ts
Comment (chrmarti) [inline src/spec-common/injectHeadless.ts:513]: Success still needs to be reported after `await Promise.all(commands)` finished successfully below. Please move it there.

**Review by chrmarti — CHANGES_REQUESTED** [review]
Left two more comments.

**Review by chrmarti — APPROVED** [review]

**Review by eljog — APPROVED** [review]

**Pull request merged**
